Assets/BossRoom/Scripts/Server/Game/Action/DashAttackAction.cs

```diff
@@ -0,0 +1,127 @@
+using MLAPI;
+using MLAPI.Spawning;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace BossRoom.Server
+{
+    /// <summary>
+    /// Causes the attacker to dash toward a target. When they stop dashing, they perform a melee attack.
+    /// The character is immune to damage while dashing.
+    ///
+    /// Since the "Range" field means "range when we can begin dashing at our target", we need another
+    /// field to mean "range of our melee attack after dashing". We'll use the "Radius" field of the
+    /// ActionDescription for that.
+    /// 
+    /// </summary>
+    /// <remarks>
+    /// See MeleeAction for relevant discussion about targeting; we use the same concept here: preferring
+    /// the chosen target, but using whatever is actually within striking distance at time of attack.
+    /// </remarks>
+    /// 
+    public class DashAttackAction : Action
+    {
+        private bool m_StartedDash;
+        private bool m_StoppedDash;
+        private float m_DashDuration;
+
+        public DashAttackAction(ServerCharacter parent, ref ActionRequestData data) : base(parent, ref data)
+        {
+            Assert.IsTrue(Description.MoveSpeed > 0, $"ActionDescription for {Description.ActionTypeEnum} needs a MoveSpeed assigned!");
+            Assert.IsTrue(Description.Radius > 0, $"ActionDescription for {Description.ActionTypeEnum} needs a Radius assigned!");
+        }
+
+        public override bool Start()
+        {
+            // snap to face our destination. This is the direction we'll charge in
+            m_Parent.transform.LookAt(GetTargetSpot());
```

Comment:
Finding some weird behaviour when this is fired when you are stationary and press the dash button on the UI. You don't have a target, so this forces you to turn towards `Data.Position` which turns out to be just your root transform's position, which will force the player to turn around and do a micro dash. I would expect the character to dash forward in the direction they're facing to some minimum distance, no? 

I also just played as the archer and this is the case too with their Attack1 skill. They turn around and shoot an arrow when stationary. Tested with a mage and this happens with their Attack1 skill as well. Seems like a shared bug with ranged attacks.

Reply:
My understanding is that if you click the action buttons it will use the last clicked location. So once you've clicked anywhere once, the attacks will aim at that spot. I'm not sure how we'd address the corner-case of not having clicked on anything yet... it would seem to me to be non-trivial for many actions (e.g. should we just find the nearest target for mage-bolt? otherwise wherever we pick, it can't possibly connect...)

Reply:
I was testing the new teleport dash and it's opened the way for speed hacks for sure. Spamming left click and 2 lets you move around the map in a very undesirable manner (I'm seeing the player get snapped to many different positions). I'm hoping `ReuseTime` can help us here. Also, is the action abiding by the range or is that set to a fairly high number? I'm able to zoom out and pick the farthest point on the map and traverse there.

Right, one solution will not work for all actions. If some actions require a target, is it worth displaying the last target position on the map, perhaps? This might help alleviate some confusion on where you're dashing, shooting at, etc.

Reply:
I've give the action a ReuseTime of 2 seconds (which can be higher if desired).

When testing this I found it to be very unfortunate that canceling the action doesn't reset the reuse time. (Meaning that even if you cancel the teleport action, you still have to wait 2 seconds before you can start again.) I've fixed `ActionPlayer` to clear the timestamp so that the action can be started again immediately if desired.

Reply:
> Also, is the action abiding by the range or is that set to a fairly high number? I'm able to zoom out and pick the farthest point on the map and traverse there.

The action's range was set pretty high at 20m (I've lowered it to 15m) but there was also a missing distance check if the action had no target. This has been fixed so that if it's too far away, we pick a new teleport spot that's closer. (As opposed to fixing it by having the character walk until they're in range, which is what happens if they have a target.)

Reply:
If the teleport-destination seems unset (is precisely Vector3.zero or is extremely close to the character's current location), we now pick a new spot that's `ActionDescription.Range` away. See what you think about this implementation.

Reply:
Much better than before. Maybe it feels a little too far when stationary but I don't object to leaving it as it is.

Reply:
Yes, i like this too.